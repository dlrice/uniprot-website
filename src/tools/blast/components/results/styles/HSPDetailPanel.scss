.hsp-detail-panel {
  height: auto;
  max-height: 50%;

  &__header {
    display: flex;
    justify-content: space-between;
    padding-top: 0.5rem;
    padding-left: 0.5rem;
    padding-right: 0.5rem;
  }
  &__visualisation {
    display: grid;
    grid-template-columns: 1fr 9fr;
    margin-bottom: 1em;
    // grid-row-gap: 1rem;
  }

  .annotation-choice:first-letter {
    text-transform: uppercase;
  }

  .hsp-locations {
    display: flex;
    flex-flow: row nowrap;
    // justify-content: space-around;
  }
}

.hsp-label {
  text-align: right;
  padding-right: 1rem;
  font-weight: 800;
  &:first-letter {
    text-transform: uppercase;
  }
}
