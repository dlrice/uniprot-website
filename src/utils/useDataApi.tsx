/* eslint-disable @typescript-eslint/no-explicit-any */
import { useState, useEffect } from 'react';
import fetchData from './fetchData';

const useDataApi = (url: string, headers: any) => {
  const [data, setData] = useState<any>({});

  useEffect(() => {
    const getData = async () => {
      const result = await fetchData(url, headers);
      setData(result.data);
    };
    getData();
  }, [url]);

  return data;
};

export default useDataApi;
