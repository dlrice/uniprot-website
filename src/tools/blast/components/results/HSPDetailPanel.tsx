/* eslint-disable no-param-reassign */
/* eslint-disable camelcase */
import React, { FC, useCallback, useState, useEffect } from 'react';
import { Loader, DropdownButton, CloseIcon } from 'franklin-sites';
import { uniq } from 'lodash-es';
import SlidingPanel from '../../../../shared/components/layouts/SlidingPanel';
import { BlastHsp } from '../../types/blastResults';
import useDataApi from '../../../../shared/hooks/useDataApi';
import { UniProtkbAPIModel } from '../../../../uniprotkb/adapters/uniProtkbConverter';
import { getAccessionsURL } from '../../../../uniprotkb/config/apiUrls';
import FeatureType from '../../../../uniprotkb/types/featureType';
import { processFeaturesData } from '../../../../uniprotkb/components/protein-data-views/FeaturesView';
import ErrorHandler from '../../../../shared/components/error-pages/ErrorHandler';
import HSPDetailOverview from './HSPDetailOverview';
import HSPDetailWrapped from './HSPDetailWrapped';
import {
  MsaColorScheme,
  msaColorSchemeToString,
} from '../../../config/msaColorSchemes';
import { findSequenceSegments } from '../../../utils';
import './styles/HSPDetailPanel.scss';

type UniProtkbAccessionsAPI = {
  results: UniProtkbAPIModel[];
};

export type HSPDetailPanelProps = {
  hsp: BlastHsp;
  hitAccession: string;
  onClose: () => void;
  hitLength: number;
  queryLength: number;
};

const HSPDetailPanel: FC<HSPDetailPanelProps> = ({
  hsp,
  hitAccession,
  onClose,
  hitLength,
  queryLength,
}) => {
  const {
    hsp_align_len,
    hsp_query_from,
    hsp_qseq,
    hsp_hit_from,
    hsp_hit_to,
    hsp_hseq,
    hsp_identity,
  } = hsp;
  // TODO calculate actual length based on total match and query lengths
  const [highlightPosition, setHighlighPosition] = useState('');
  const [annotation, setAnnotation] = useState<FeatureType>();
  const [highlightProperty, setHighlightProperty] = useState<MsaColorScheme>();
  const [initialDisplayEnd, setInitialDisplayEnd] = useState<
    number | undefined
  >();
  const [activeView, setActiveView] = useState<'overview' | 'wrapped'>(
    'overview'
  );
  const tracksOffset = Math.abs(hsp_hit_from - hsp_query_from);

  useEffect(() => {
    setHighlighPosition('0:0');
  }, [hsp_hit_from, hsp_query_from]);

  const setQueryTrackData = useCallback(
    (node): void => {
      if (node) {
        const seqSegments = findSequenceSegments(hsp_qseq);
        const opacity = hsp_identity / 100;

        const blockSegments = seqSegments.map(([start, end]) => ({
          start: start + tracksOffset,
          end: end + tracksOffset,
          // franklin $colour-sapphire-blue
          color: '#014371',
          opacity,
        }));

        const lineSegments = [
          {
            start: tracksOffset + 1,
            end: hsp_query_from + 1,
            shape: 'line',
            color: '#014371',
            opacity,
          },
          {
            start: blockSegments[blockSegments.length - 1].end + 1,
            end: queryLength + tracksOffset,
            shape: 'line',
            color: '#014371',
            opacity,
          },
        ];

        // eslint-disable-next-line no-param-reassign
        node.data = [...lineSegments, ...blockSegments];
      }
    },
    [hsp_query_from, hsp_identity, hsp_qseq, queryLength, tracksOffset]
  );

  const setMatchTrackData = useCallback(
    (node): void => {
      if (node) {
        const seqSegments = findSequenceSegments(hsp_hseq);
        const opacity = hsp_identity / 100;

        const blockSegments = seqSegments.map(([start, end]) => ({
          start: start + tracksOffset,
          end: end + tracksOffset,
          // franklin $colour-sapphire-blue
          color: '#014371',
          opacity,
        }));

        const lineSegments = [
          {
            start: 1,
            end: tracksOffset,
            shape: 'line',
            color: '#014371',
            opacity,
          },
          {
            start: blockSegments[blockSegments.length - 1].end + 1,
            end: hitLength,
            shape: 'line',
            color: '#014371',
            opacity,
          },
        ];

        // eslint-disable-next-line no-param-reassign
        node.data = [...lineSegments, ...blockSegments];
      }
    },
    [hsp_identity, hitLength, hsp_hseq, tracksOffset]
  );

  const setMSAAttributes = useCallback(
    (node): void => {
      if (!node) {
        return;
      }
      setInitialDisplayEnd(hsp_align_len / (15 / node.getSingleBaseWidth()));
      node.data = [
        {
          name: 'Query',
          sequence: hsp_qseq,
        },
        {
          name: 'Match',
          sequence: hsp_hseq,
        },
      ];
    },
    [hsp_qseq, hsp_hseq, hsp_align_len]
  );

  const { loading, data, status, error } = useDataApi<UniProtkbAccessionsAPI>(
    getAccessionsURL([hitAccession], { facets: [] })
  );

  const features = data?.results?.[0]?.features;
  const recommendedName =
    data?.results?.[0]?.proteinDescription?.recommendedName?.fullName.value;
  const organism = data?.results?.[0]?.organism?.scientificName;

  const title = [hitAccession, recommendedName, organism]
    .filter(Boolean)
    .join(' Â· ');

  const annotationChoices = uniq(features?.map(({ type }) => type));
  if (!annotation && !!annotationChoices?.length) {
    setAnnotation(annotationChoices[0]);
  }

  const setFeatureTrackData = useCallback(
    (node): void => {
      if (node && features && annotation) {
        let processedFeatures = processFeaturesData(
          features.filter(({ type }) => type === annotation)
        );
        if (activeView === 'wrapped') {
          processedFeatures = processedFeatures
            .map((feature) => ({
              ...feature,
              start: feature.start - hsp_hit_from,
            }))
            .filter(({ start }) => start >= 0 && start <= hsp_hit_to);
        }
        node.data = processedFeatures;
      }
    },
    [features, annotation, activeView, hsp_hit_from, hsp_hit_to]
  );

  type EventDetail = {
    displaystart: string;
    displayend: string;
  };

  const findHighlighPositions = useCallback(
    ({ displaystart, displayend }: EventDetail) => {
      const start = tracksOffset + parseInt(displaystart, 10);
      const end = tracksOffset + parseInt(displayend, 10);
      setHighlighPosition(`${start}:${end}`);
    },
    [tracksOffset]
  );

  const managerRef = useCallback(
    (node): void => {
      if (node && initialDisplayEnd) {
        node.addEventListener('change', ({ detail }: { detail: EventDetail }) =>
          findHighlighPositions(detail)
        );
        node.setAttribute('displaystart', 1);
        node.setAttribute('displayend', initialDisplayEnd);
        setHighlighPosition(
          `${tracksOffset + 1}:${tracksOffset + initialDisplayEnd}`
        );
      }
    },
    [initialDisplayEnd, findHighlighPositions, tracksOffset]
  );

  if (error) {
    return <ErrorHandler status={status} />;
  }

  if (loading) {
    return <Loader />;
  }
  const conservationOptions =
    highlightProperty === MsaColorScheme.CONSERVATION
      ? {
          'calculate-conservation': true,
          'overlay-conservation': true,
        }
      : {};

  return (
    <SlidingPanel position="bottom" className="hsp-detail-panel">
      <div className="hsp-detail-panel__header">
        <h4>{title}</h4>
        <button type="button" onClick={onClose}>
          <CloseIcon width="16" height="16" />
        </button>
      </div>

      <div className="button-group">
        <DropdownButton label="Highlight properties" className="tertiary">
          {(setShowMenu: (showMenu: boolean) => void) => (
            <div className="dropdown-menu__content">
              <ul>
                {Object.entries(msaColorSchemeToString).map(
                  ([schemeValue, schemeString]) => (
                    // TODO: indicate currently selected
                    <li key={schemeString}>
                      <button
                        type="button"
                        onClick={() => {
                          setShowMenu(false);
                          setHighlightProperty(schemeValue as MsaColorScheme);
                        }}
                      >
                        {schemeString}
                      </button>
                    </li>
                  )
                )}
              </ul>
            </div>
          )}
        </DropdownButton>
        {!!annotationChoices?.length && (
          <DropdownButton label="Show annotation" className="tertiary">
            {(setShowMenu: (showMenu: boolean) => void) => (
              <div className="dropdown-menu__content">
                <ul>
                  {annotationChoices.map((annotationChoice) => (
                    // TODO: indicate currently selected
                    <li key={annotationChoice}>
                      <button
                        type="button"
                        className="annotation-choice"
                        onClick={() => {
                          setShowMenu(false);
                          setAnnotation(annotationChoice);
                        }}
                      >
                        {annotationChoice}
                      </button>
                    </li>
                  ))}
                </ul>
              </div>
            )}
          </DropdownButton>
        )}
        <DropdownButton label="View" className="tertiary">
          {(setShowMenu: (showMenu: boolean) => void) => (
            <div className="dropdown-menu__content">
              <ul>
                <li key="overview">
                  <button
                    type="button"
                    className="button tertiary"
                    onClick={() => {
                      setShowMenu(false);
                      setActiveView('overview');
                    }}
                  >
                    Overview
                  </button>
                </li>
                <li key="wrapped">
                  <button
                    type="button"
                    className="button tertiary"
                    onClick={() => {
                      setShowMenu(false);
                      setActiveView('wrapped');
                    }}
                  >
                    Wrapped
                  </button>
                </li>
              </ul>
            </div>
          )}
        </DropdownButton>
      </div>
      <div className="hsp-detail-panel__body">
        {activeView === 'overview' ? (
          <HSPDetailOverview
            managerRef={managerRef}
            hsp_align_len={hsp_align_len}
            setMSAAttributes={setMSAAttributes}
            highlightProperty={highlightProperty}
            conservationOptions={conservationOptions}
            setQueryTrackData={setQueryTrackData}
            hitLength={hitLength}
            highlightPosition={highlightPosition}
            hitAccession={hitAccession}
            setMatchTrackData={setMatchTrackData}
            annotation={annotation}
            setFeatureTrackData={setFeatureTrackData}
          />
        ) : (
          <HSPDetailWrapped
            managerRef={managerRef}
            hsp_align_len={hsp_align_len}
            hsp_qseq={hsp_qseq}
            hsp_hseq={hsp_hseq}
            hsp_hit_from={hsp_hit_from}
            hsp_query_from={hsp_query_from}
            highlightProperty={highlightProperty}
            conservationOptions={conservationOptions}
            setQueryTrackData={setQueryTrackData}
            highlightPosition={highlightPosition}
            hitAccession={hitAccession}
            setMatchTrackData={setMatchTrackData}
            annotation={annotation}
            setFeatureTrackData={setFeatureTrackData}
          />
        )}
      </div>
    </SlidingPanel>
  );
};

export default HSPDetailPanel;
