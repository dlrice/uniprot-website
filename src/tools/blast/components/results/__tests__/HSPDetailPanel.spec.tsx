import React from 'react';
import { fireEvent } from '@testing-library/react';
import blastResultsMockData from '../../../../__mocks__/server-jobs/example-truncated.json';
import HSPDetailPanel, { transformFeaturesPositions } from '../HSPDetailPanel';
import modelData from '../../../../../uniprotkb/__mocks__/entryModelData.json';
import renderWithRouter from '../../../../../shared/__test-helpers__/RenderWithRouter';

jest.mock('../../../../../shared/hooks/useDataApi', () => jest.fn());
import useDataApi from '../../../../../shared/hooks/useDataApi';

const dataMock = {
  loading: false,
  data: modelData,
};
useDataApi.mockImplementation(() => dataMock);

describe('HSPDetailPanel', () => {
  let rendered;
  const onClose = jest.fn();
  const hit = blastResultsMockData.hits[0];
  const hsp = hit.hit_hsps[0];

  beforeEach(async () => {
    rendered = renderWithRouter(
      <HSPDetailPanel
        hsp={hsp}
        hitAccession={hit.hit_acc}
        onClose={onClose}
        hitLength={hit.hit_len}
        queryLength={blastResultsMockData.query_len}
      />
    );
    await rendered;
  });

  it('should initially render overview', () => {
    const { asFragment, getByTestId } = rendered;
    expect(getByTestId('overview-hsp-detail')).toBeTruthy();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should load correct query and match sequence data', async () => {
    const { container } = rendered;
    const msa = container.querySelector('protvista-msa');
    expect(msa.data).toEqual([
      { name: 'Query', sequence: hsp.hsp_qseq },
      { name: 'Match', sequence: hsp.hsp_hseq },
    ]);
  });

  it('should change to wrapped and render when wrapped view is clicked', async () => {
    const { getByText, getByTestId, asFragment } = rendered;
    const wrappedButton = getByText('Wrapped');
    fireEvent.click(wrappedButton);
    expect(getByTestId('wrapped-hsp-detail')).toBeTruthy();
    expect(asFragment()).toMatchSnapshot();
  });
});
