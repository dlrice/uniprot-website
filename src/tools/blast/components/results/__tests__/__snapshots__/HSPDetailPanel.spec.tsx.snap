// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HSPDetailPanel should change to wrapped and render when wrapped view is clicked 1`] = `
<div>
  <div
    class="hsp-detail-panel__header"
  >
    <h4>
      U6MKN0 路 rec full Name 路 scientific name
    </h4>
    <button
      type="button"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 352 512"
        width="16"
      >
        <path
          d="M243 256l100-100c12-12 12-32 0-45l-22-22a31 31 0 00-45 0L176 189 76 89a31 31 0 00-45 0L9 111a31 31 0 000 45l100 100L9 356a31 31 0 000 45l22 22c13 12 33 12 45 0l100-100 100 100c12 12 32 12 45 0l22-22c12-13 12-33 0-45L243 256z"
        />
      </svg>
    </button>
  </div>
  <div
    class="hsp-detail-panel__body"
  >
    <div
      class="button-group"
    >
      <div
        class="dropdown-container"
      >
        <button
          class="button dropdown tertiary"
          type="button"
        >
          Highlight properties
        </button>
        <div
          class="dropdown-menu"
        >
          <div
            class="dropdown-menu__content"
          >
            <ul>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Similarity
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Negative
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Positive
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Aliphatic
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Aromatic
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Charged
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Polar
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Serine Threonine
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Buried Index
                </button>
              </li>
              <li>
                <button
                  class="button primary"
                  type="button"
                >
                  Clustal
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Clustal 2
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Cinema
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Helix Propensity
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Hydro
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Lesk
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Mae
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Nucleotide
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Purine Pyrimidine
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Strand Propensity
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Taylor
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Turn Propensity
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Zappo
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="dropdown-container"
      >
        <button
          class="button dropdown tertiary"
          type="button"
        >
          Show annotation
        </button>
        <div
          class="dropdown-menu"
        >
          <div
            class="dropdown-menu__content"
          >
            <ul>
              <li>
                <button
                  class="button annotation-choice primary"
                  type="button"
                >
                  chain
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <fieldset
        class="msa-view-choice"
      >
        View:
        <label>
          <input
            checked=""
            name="view"
            type="radio"
          />
          Overview
        </label>
        <label>
          <input
            name="view"
            type="radio"
          />
          Wrapped
        </label>
      </fieldset>
    </div>
    <div>
      <div>
        <section
          class="alignment-grid"
          data-testid="alignment-wrapped-view"
        >
          <section
            class="alignment-grid__row alignment-grid__row--msa-track"
          >
            <div
              class="track-label track-label--align-labels"
            >
              <span
                class="align-label"
              >
                Query
              </span>
              <span
                class="align-label"
                style="height: 20px;"
              >
                Match
                <a
                  href="/uniprotkb/U6MKN0"
                >
                  U6MKN0
                </a>
                <button
                  class="button tertiary"
                  type="button"
                />
              </span>
            </div>
            <div
              class="track"
            >
              <protvista-msa
                colorscheme="clustal"
                height="40"
                length="45"
              />
            </div>
            <span
              class="right-coord"
            >
              <div
                style="height: 20px; margin-top: 0px;"
              >
                47
              </div>
              <div
                style="height: 20px;"
              >
                462
              </div>
            </span>
          </section>
          <section
            class="alignment-grid__row"
          >
            <span
              class="track-label"
            >
              chain
            </span>
            <div
              class="track"
            >
              <protvista-track
                displayend="462"
                displaystart="418"
                length="45"
              />
            </div>
          </section>
        </section>
      </div>
    </div>
  </div>
</div>
`;

exports[`HSPDetailPanel should initially render overview 1`] = `
<DocumentFragment>
  <div
    class="sliding-panel sliding-panel--bottom hsp-detail-panel"
    style="opacity: 0; margin-right: -1000px;"
  >
    <div>
      <div
        class="hsp-detail-panel__header"
      >
        <h4>
          U6MKN0 路 rec full Name 路 scientific name
        </h4>
        <button
          type="button"
        >
          <svg
            fill="currentColor"
            height="16"
            viewBox="0 0 352 512"
            width="16"
          >
            <path
              d="M243 256l100-100c12-12 12-32 0-45l-22-22a31 31 0 00-45 0L176 189 76 89a31 31 0 00-45 0L9 111a31 31 0 000 45l100 100L9 356a31 31 0 000 45l22 22c13 12 33 12 45 0l100-100 100 100c12 12 32 12 45 0l22-22c12-13 12-33 0-45L243 256z"
            />
          </svg>
        </button>
      </div>
      <div
        class="hsp-detail-panel__body"
      >
        <div
          class="button-group"
        >
          <div
            class="dropdown-container"
          >
            <button
              class="button dropdown tertiary"
              type="button"
            >
              Highlight properties
            </button>
            <div
              class="dropdown-menu"
            >
              <div
                class="dropdown-menu__content"
              >
                <ul>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Similarity
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Negative
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Positive
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Aliphatic
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Aromatic
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Charged
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Polar
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Serine Threonine
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Buried Index
                    </button>
                  </li>
                  <li>
                    <button
                      class="button primary"
                      type="button"
                    >
                      Clustal
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Clustal 2
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Cinema
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Helix Propensity
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Hydro
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Lesk
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Mae
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Nucleotide
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Purine Pyrimidine
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Strand Propensity
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Taylor
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Turn Propensity
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Zappo
                    </button>
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <div
            class="dropdown-container"
          >
            <button
              class="button dropdown tertiary"
              type="button"
            >
              Show annotation
            </button>
            <div
              class="dropdown-menu"
            >
              <div
                class="dropdown-menu__content"
              >
                <ul>
                  <li>
                    <button
                      class="button annotation-choice primary"
                      type="button"
                    >
                      chain
                    </button>
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <fieldset
            class="msa-view-choice"
          >
            View:
            <label>
              <input
                checked=""
                name="view"
                type="radio"
              />
              Overview
            </label>
            <label>
              <input
                name="view"
                type="radio"
              />
              Wrapped
            </label>
          </fieldset>
        </div>
        <div>
          <section
            class="alignment-grid"
            data-testid="alignment-view"
          >
            <section
              class="alignment-grid__row"
            >
              <span
                class="track-label"
              >
                Overview
              </span>
              <div
                class="track"
              >
                <div>
                  <protvista-track
                    height="15"
                    highlight="418:446"
                    layout="non-overlapping"
                    length="486"
                  />
                  <protvista-track
                    height="15"
                    highlight="418:446"
                    layout="non-overlapping"
                    length="486"
                  />
                </div>
              </div>
            </section>
            <section
              class="alignment-grid__row"
            >
              <span
                class="track-label"
              >
                chain
              </span>
              <div
                class="track"
              >
                <protvista-track
                  highlight="418:446"
                  layout="non-overlapping"
                  length="486"
                />
              </div>
            </section>
            <section
              class="alignment-grid__row alignment-grid__row--msa-track"
            >
              <div
                class="track-label track-label--align-labels"
              >
                <span
                  class="align-label"
                >
                  Query
                </span>
                <span
                  class="align-label"
                  style="height: 20px;"
                >
                  Match
                  <a
                    href="/uniprotkb/U6MKN0"
                  >
                    U6MKN0
                  </a>
                  <button
                    class="button tertiary"
                    type="button"
                  />
                </span>
              </div>
              <div
                class="track"
              >
                <protvista-manager
                  attributes="displaystart displayend"
                  displayend="28"
                  displaystart="1"
                >
                  <protvista-navigation
                    length="28"
                  />
                  <protvista-msa
                    colorscheme="clustal"
                    height="40"
                    length="28"
                  />
                </protvista-manager>
              </div>
              <span
                class="right-coord"
              >
                <div
                  style="height: 20px; margin-top: 0px;"
                >
                  28
                </div>
                <div
                  style="height: 20px;"
                >
                  28
                </div>
              </span>
            </section>
          </section>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
